File: src/backend/Domain/AgentPlan.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace MoraKnode.Domain;

/// <summary>
/// ADR-002: every agent-driven plan goes through a human (or human-class)
/// approval gate before downstream work happens. One AgentPlan = one
/// proposal for how to attack a Task. Hybrid shape (Steps + free-text Notes)
/// because rigid step lists feel forced for some kinds of work but a single
/// blob loses the structured-approval feel.
///
/// PendingReview → Approved / Rejected is terminal. A revision = submit a
/// new AgentPlan against the same TaskId; the older row stays in place for
/// audit. v1 has no soft-edit on PendingReview either — to revise an
/// in-flight plan, the agent submits a new one (and the previous remains
/// visible in the task's plan history).
/// </summary>
public class AgentPlan
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonRepresentation(BsonType.ObjectId)]
    public string TaskId { get; set; } = null!;

    /// <summary>The Resource that submitted this plan — typically an agent,
    /// but a human can also submit (so the same gate works for human-authored
    /// proposals when policy demands review).</summary>
    [BsonRepresentation(BsonType.ObjectId)]
    public string SubmittedByResourceId { get; set; } = null!;

    [BsonRepresentation(BsonType.String)]
    public PlanStatus Status { get; set; } = PlanStatus.PendingReview;

    public string Title { get; set; } = string.Empty;
    public int? EstimateMinutes { get; set; }

    public List<PlanStep> Steps { get; set; } = new();
    public string? Notes { get; set; }

    /// <summary>Reviewer's note on approve or (especially) reject. Required
    /// on reject so the agent knows what to fix when re-submitting.</summary>
    public string? ReviewerComment { get; set; }

    [BsonRepresentation(BsonType.ObjectId)]
    public string? ReviewedByResourceId { get; set; }

    public DateTime? ReviewedAt { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

public class PlanStep
{
    public string Description { get; set; } = string.Empty;
    public int? EstimateMinutes { get; set; }
}

public enum PlanStatus
{
    PendingReview,
    Approved,
    Rejected
}

File: src/frontend/lib/models/agent_plan.dart
/// Mirrors backend Domain/AgentPlan.cs PlanStatus enum. Backend serializes
/// the enum by name (BsonRepresentation.String) so the wire format is
/// human-readable across migrations.
enum PlanStatus {
  pendingReview,
  approved,
  rejected;

  static PlanStatus parse(Object? v) {
    final n = v?.toString().toLowerCase();
    return switch (n) {
      'approved' => PlanStatus.approved,
      'rejected' => PlanStatus.rejected,
      _ => PlanStatus.pendingReview,
    };
  }

  String get wire => switch (this) {
        PlanStatus.pendingReview => 'PendingReview',
        PlanStatus.approved => 'Approved',
        PlanStatus.rejected => 'Rejected',
      };
}

/// One step inside a plan. Free-form description + optional minute estimate.
/// The backend stores PlanStep as a sub-document inside AgentPlan.Steps.
class PlanStep {
  final String description;
  final int? estimateMinutes;

  const PlanStep({required this.description, this.estimateMinutes});

  factory PlanStep.fromJson(Map<String, dynamic> json) => PlanStep(
        description: (json['description'] as String?) ?? '',
        estimateMinutes: (json['estimateMinutes'] as num?)?.toInt(),
      );
}

/// Mirrors Domain/AgentPlan.cs. PendingReview plans are what the review
/// queue surfaces; Approved / Rejected are terminal and shown in the
/// per-task plan history.
class AgentPlan {
  final String id;
  final String taskId;
  final String submittedByResourceId;
  final PlanStatus status;
  final String title;
  final int? estimateMinutes;
  final List<PlanStep> steps;
  final String? notes;
  final String? reviewerComment;
  final String? reviewedByResourceId;
  final DateTime? reviewedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const AgentPlan({
    required this.id,
    required this.taskId,
    required this.submittedByResourceId,
    required this.status,
    required this.title,
    this.estimateMinutes,
    this.steps = const [],
    this.notes,
    this.reviewerComment,
    this.reviewedByResourceId,
    this.reviewedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory AgentPlan.fromJson(Map<String, dynamic> json) => AgentPlan(
        id: json['id'] as String,
        taskId: json['taskId'] as String,
        submittedByResourceId: json['submittedByResourceId'] as String,
        status: PlanStatus.parse(json['status']),
        title: (json['title'] as String?) ?? '',
        estimateMinutes: (json['estimateMinutes'] as num?)?.toInt(),
        steps: ((json['steps'] as List?) ?? const [])
            .cast<Map<String, dynamic>>()
            .map(PlanStep.fromJson)
            .toList(),
        notes: json['notes'] as String?,
        reviewerComment: json['reviewerComment'] as String?,
        reviewedByResourceId: json['reviewedByResourceId'] as String?,
        reviewedAt: _parseDate(json['reviewedAt']),
        createdAt: DateTime.parse(json['createdAt'] as String).toUtc(),
        updatedAt: DateTime.parse(json['updatedAt'] as String).toUtc(),
      );

  static DateTime? _parseDate(Object? v) =>
      v == null ? null : DateTime.parse(v as String).toUtc();
}
